### PR TITLE
Prepare C-language infrastructure for ISO:SQL

### PR DESCRIPTION
**Pull request by mkgrgis**
No SQL behaviour changes.

Preparing to implement SQL behaviour from table discussed in https://github.com/pgspider/sqlite_fdw/issues/66#issuecomment-1504830541

This PR include:
* improve naming;
* add function for human readable error message about SQLite data value affinity;
* change return type for `sqlite_convert_to_pg` from `Datum` to `NullableDatum` for future proper `uuid` processing and other special cases such as void `text` affinity in data for `int` column;
* add TODO comments about PostgreSQL database encoding in proper places;
* reduce double calculating for two functions: `sqlite3_column_bytes` and `sqlite3_column_bytes` for the same data value;
* add draft for human readable error message about SQLite data value affinity;
* etc.

### CODE REVIEW

File: sqlite_fdw.c [hunk 1]
```diff
@@ -1572,14 +1573,36 @@ make_tuple_from_result_row(sqlite3_stmt * stmt,
...
 
-		if (sqlite3_column_type(stmt, attid) != SQLITE_NULL)
+		sqlite_value_affinity = sqlite3_column_type(stmt, stmt_colid);
+		if ( sqlite_value_affinity != SQLITE_NULL)
 		{
-			is_null[attnum] = false;
-			row[attnum] = sqlite_convert_to_pg(pgtype, pgtypmod,
-											   stmt, attid, festate->attinmeta, attnum);
+		    /* Here will be processing of column options about special convert behaviour 
+		    options = GetForeignColumnOptions(rel, attnum_base);
```
Comment (nxhai98) [inline sqlite_fdw.c:1583]: Is this code unnecessary? If so, please remove it.

File: sqlite_query.c [hunk 2]
```diff
@@ -21,63 +21,82 @@
...
 		case BYTEAOID:
 			{
-				int			blobsize = sqlite3_column_bytes(stmt, stmt_colid);
-
-				value_datum = (Datum) palloc0(blobsize + VARHDRSZ);
-				memcpy(VARDATA(value_datum), sqlite3_column_blob(stmt, stmt_colid), blobsize);
-				SET_VARSIZE(value_datum, blobsize + VARHDRSZ);
-				return PointerGetDatum(value_datum);
+				// int			value_byte_size_blob_or_utf = sqlite3_column_bytes(stmt, stmt_colid); // Calculated always for detectind void values
+				value_datum = (Datum) palloc0(value_byte_size_blob_or_utf8 + VARHDRSZ);
```
Comment (nxhai98) [inline sqlite_query.c:66]: `value_byte_size_blob_or_utf8` is used only in this block, could you explain why we need define it outside?  
I do not understand your comment as well: `// Compute always, void text and void BLOB will be special cases`

File: sqlite_query.c [hunk 3]
```diff
@@ -111,15 +131,24 @@ sqlite_convert_to_pg(Oid pgtyp, int pgtypmod, sqlite3_stmt * stmt, int stmt_coli
 				valstr = DatumGetCString(DirectFunctionCall1(float8out, Float8GetDatum((float8) value)));
 				break;
 			}
+		/* some popular datatypes for default algorythm branch
```
Comment (nxhai98) [inline sqlite_query.c:134]: Please using comment style:
```
/*
 * some comment
 */
```

**Review by nxhai98 — COMMENTED** [review]
I have some comments for your PR. Could you check it?

**Comment by mkgrgis** [comment]
Thanks, @nxhai98 -san! I have fixed some problems in https://github.com/mkgrgis/sqlite_fdw/commit/74bcfeaf61fd0b5db579995e4a7fcd7cb8263a08

File: deparse.c [hunk 4]
```diff
@@ -2092,6 +2092,7 @@ sqlite_deparse_column_option(int varno, int varattno, PlannerInfo *root, char *o
 void
 sqlite_deparse_string_literal(StringInfo buf, const char *val)
 {
+    // TODO: text input for SQLite is always UTF-8, we need to respect PostgreSQL database encoding
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply deparse.c]: Fixed in https://github.com/mkgrgis/sqlite_fdw/tree/convert_data_types, also added
https://github.com/mkgrgis/sqlite_fdw/commit/7262db8fc3ea3fb3ae0700a22084ed8ba3c2a680 about C code style.

File: sqlite_fdw.c (hunk 1 above)
Comment (mkgrgis) [reply sqlite_fdw.c:1583]: This code will be uncommented during implementation of [data transformation table](https://github.com/mkgrgis/sqlite_fdw/tree/readme-fix-and-additions#datatypes). @t-kataym have wrote this implementation need to change SQL behaviour during column and other levelled options. It's important to leave this necessary draft  for other contributors. My next PR will be about PostgreSQL encoding, PR after next about human readable error message (uncommenting) and than first PR with [data transformation table](https://github.com/mkgrgis/sqlite_fdw/tree/readme-fix-and-additions#datatypes) implementation and introducing a new option or options.

File: sqlite_query.c [hunk 5]
```diff
@@ -21,63 +21,82 @@
...
-	int			col_type = sqlite3_column_type(stmt, stmt_colid);
-	int			sqlite_type = sqlite_from_pgtyp(pgtyp);
+	int			affinity_for_pg_column = sqlite_affinity_eqv_to_pgtype(pgtyp);
+	int         value_byte_size_blob_or_utf8 = sqlite3_column_bytes(stmt, stmt_colid); // Compute always, void text and void BLOB will be special cases
 
-	if (sqlite_type != col_type && col_type == SQLITE3_TEXT)
-		elog(ERROR, "invalid input syntax for type =%d, column type =%d", sqlite_type, col_type);
+	if (affinity_for_pg_column != sqlite_value_affinity && sqlite_value_affinity == SQLITE3_TEXT)
+	{
+	    /* Here will be human readable error message
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply sqlite_query.c]: My next PR will be about PostgreSQL encoding, PR after next about human readable error message (uncommenting). So, this code will be uncommented in a small  PR after next small PR if this PR will be accepted.

File: sqlite_query.c (hunk 2 above)
Comment (mkgrgis) [reply sqlite_query.c:66]: In our  [data transformation table](https://github.com/mkgrgis/sqlite_fdw/tree/readme-fix-and-additions#datatypes) empty value with `text` affinity is special case. For example for PostgreSQL `uuid` we *should* always threat void `text` value as `NULL`. For PostgreSQL `int` we can threat this value as `NULL` if a special boolean option is `true`. So for all PostgreSQL data types `if (value_byte_size_blob_or_utf8)` will be necessary.

File: sqlite_query.c (hunk 3 above)
Comment (mkgrgis) [reply sqlite_query.c]: Fixed, please verify.

File: sqlite_fdw.c (hunk 1 above)
Comment (mkgrgis) [reply sqlite_fdw.c:1583]: I don't want to break any tests with the message in this PR, but will uncomment in near future.

File: sqlite_query.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkgrgis) [reply sqlite_query.c]: I don't want to break any tests with the message in this PR, but will uncomment in near future.

File: deparse.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply deparse.c]: I checked the change, but it not same as my expectation. Could you refer this patch: 
[fix_comment_style.patch](https://github.com/pgspider/sqlite_fdw/files/11474843/fix_comment_style.patch)

File: sqlite_fdw.c (hunk 1 above)
Comment (nxhai98) [reply sqlite_fdw.c:1583]: Thank for your explanation,

However, we cannot accept the PR containing such comment code.
Leaving commented-out code in source code is bad practice, as it takes up space, causes confusion, and leads to maintenance issues if it's not removed.

Could you remove it or change it to a TODO note?

File: sqlite_query.c (hunk 2 above)
Comment (nxhai98) [reply sqlite_query.c:66]: Thank for your explanation.   
I understood.

File: sqlite_fdw.c (hunk 1 above)
Comment (mkgrgis) [reply sqlite_fdw.c]: Yes, I have reduced this draft to one TODO comment line. It's really "leaving commented-out code in source code is bad practice".

File: sqlite_query.c (hunk 3 above)
Comment (mkgrgis) [reply sqlite_query.c:134]: Fixed again, @nxhai98 !

File: sqlite_fdw.c (hunk 1 above)
Comment (nxhai98) [reply sqlite_fdw.c]: Could you check the other place also?

Comment (mkgrgis) [reply sqlite_fdw.c]: Yes, please comment on lines after my [new commit](https://github.com/pgspider/sqlite_fdw/pull/74/commits/90d9064a4195be51214dd34088541a202a64a9af) with missing code style. Now I have no idea where is problems.

File: deparse.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkgrgis) [reply deparse.c]: I think now all my comments are /t aligned and 2nd+ lines also begins from some /t and one whitespace. Please show a problems not fixed by me.

File: sqlite_fdw.c [hunk 6]
```diff
@@ -1572,14 +1573,33 @@ make_tuple_from_result_row(sqlite3_stmt * stmt,
 		int			attnum = lfirst_int(lc) - 1;
 		Oid			pgtype = TupleDescAttr(tupleDescriptor, attnum)->atttypid;
 		int32		pgtypmod = TupleDescAttr(tupleDescriptor, attnum)->atttypmod;
+		//List	   *options = NULL;
```
Comment (nxhai98) [inline sqlite_fdw.c:1576]: Please remove it.

File: sqlite_fdw.c [hunk 7]
```diff
@@ -1572,14 +1573,33 @@ make_tuple_from_result_row(sqlite3_stmt * stmt,
...
+		int			sqlite_value_affinity;
 
-		if (sqlite3_column_type(stmt, attid) != SQLITE_NULL)
+		sqlite_value_affinity = sqlite3_column_type(stmt, stmt_colid);
+		if ( sqlite_value_affinity != SQLITE_NULL)
 		{
-			is_null[attnum] = false;
-			row[attnum] = sqlite_convert_to_pg(pgtype, pgtypmod,
-											   stmt, attid, festate->attinmeta, attnum);
+		    /* TODO: Processing of column options about special convert behaviour
```
Comment (nxhai98) [inline sqlite_fdw.c:1582]: This indent is not correct. Please using \t instead of 3 spaces

File: sqlite_query.c [hunk 8]
```diff
@@ -21,63 +21,82 @@
...
-	int			col_type = sqlite3_column_type(stmt, stmt_colid);
-	int			sqlite_type = sqlite_from_pgtyp(pgtyp);
+	int			affinity_for_pg_column = sqlite_affinity_eqv_to_pgtype(pgtyp);
+	int		 	value_byte_size_blob_or_utf8 = sqlite3_column_bytes(stmt, stmt_colid); // Compute always, void text and void BLOB will be special cases
 
-	if (sqlite_type != col_type && col_type == SQLITE3_TEXT)
-		elog(ERROR, "invalid input syntax for type =%d, column type =%d", sqlite_type, col_type);
+	if (affinity_for_pg_column != sqlite_value_affinity && sqlite_value_affinity == SQLITE3_TEXT)
+	{
+		/* TODO: human readable error message based on this code
```
Comment (nxhai98) [inline sqlite_query.c:46]: I think the error message format or error message example is enough, not need to add the code here.

File: sqlite_query.c [hunk 9]
```diff
@@ -111,15 +131,27 @@ sqlite_convert_to_pg(Oid pgtyp, int pgtypmod, sqlite3_stmt * stmt, int stmt_coli
...
+		 * case BPCHAROID:
+		 * case VARCHAROID:
+		 * case TEXTOID:
+		 * case JSONOID:
+		 * case NAMEOID:
+		 * case TIMEOID:
+		 */
 		default:
-			valstr = (char *) sqlite3_column_text(stmt, stmt_colid);
+			{	/*
```
Comment (nxhai98) [inline sqlite_query.c:143]: Please enter the comment to the new line.

File: sqlite_fdw.c [hunk 10]
```diff
@@ -1572,14 +1573,33 @@ make_tuple_from_result_row(sqlite3_stmt * stmt,
...
+		    /* TODO: Processing of column options about special convert behaviour 
+		     * options = GetForeignColumnOptions(rel, attnum_base); ... foreach(lc_attr, options)
+			 */
+			
+			int AffinityBehaviourFlags = 0;
+			/* TODO
+			 * Flags about special convert behaviour from options on database, table or column level
+			 */
+
+            sqlite_coverted = sqlite_convert_to_pg(pgtype, pgtypmod,
```
Comment (nxhai98) [inline sqlite_fdw.c:1591]: Please using \t instead of spaces.

File: sqlite_query.c (hunk 3 above)
Comment (nxhai98) [reply sqlite_query.c:134]: Thank you.

**Review by nxhai98 — COMMENTED** [review]
I add some more comments. Could you please help me check it again?

File: sqlite_fdw.c (hunk 6 above)
Comment (mkgrgis) [reply sqlite_fdw.c:1576]: Removed

File: sqlite_query.c (hunk 8 above)
Comment (mkgrgis) [reply sqlite_query.c:46]: Really. I have reduced the comment.

File: sqlite_query.c (hunk 9 above)
Comment (mkgrgis) [reply sqlite_query.c:143]: Changed

File: sqlite_fdw.c (hunk 10 above)
Comment (mkgrgis) [reply sqlite_fdw.c:1591]: Here and near fixed

**Comment by mkgrgis** [comment]
> I add some more comments. Could you please help me check it again?

Yes. Thanks, @nxhai98 !

File: deparse.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply deparse.c]: Thank you.

File: sqlite_query.c (hunk 8 above)
Comment (nxhai98) [reply sqlite_query.c:46]: Thank you.

**Review by nxhai98 — APPROVED** [review]
Thank for your fixing. I confirmed all the change.

File: sqlite_fdw.c (hunk 6 above)
Comment (nxhai98) [reply sqlite_fdw.c]: Thank you.

File: sqlite_query.c (hunk 9 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank you.

File: sqlite_fdw.c (hunk 10 above)
Comment (nxhai98) [reply sqlite_fdw.c]: Thank you.

File: sqlite_fdw.c (hunk 7 above)
Comment (mkgrgis) [reply sqlite_fdw.c]: Fixed

**Comment by mkgrgis** [comment]
@t-kataym , yes, you can squash and merge my PR. Reviews by @nxhai98 was very good and helpful for me! My next little PR will be about database encoding, also without breaking of any tests.

**Comment by t-kataym** [comment]
@mkgrgis , Thank you for your reply and development. I will merge it.

**Pull request merged**